### DISCUSSION TITLE
Bug Report: SB Keywords v3.0 & v3.2 list returns incorrectly rounded bigint campaignId & adGroupId

### DISCUSSION THREAD

**Discussion by ncoughlin** [amzn/ads-advanced-tools-docs#81] [Q&A] [answered]
I'm pretty sure I've stumbled onto a bug in the Sponsored Brands Keywords v3.0 & v3.2 API.

Regarding this list keywords endpoint:  https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi#/Keywords/listKeywords

The adGroupId's and CampaignId's that are returned in the list are incorrect, but only for the ones that would be of **bigint** datatype in the database.

I believe that is the case because the id's that are returned are very close to valid campaign and ad group id's, but with the last 2-3 digits being incorrect, which is the exact same error that I have encountered recently while exporting my own data from Athena. The error is most likely occurring while converting a bigint data type to a javascript/json number, as the id's exceed the [Number.MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER). The conversion is incorrectly rounding the numbers, making the id's useless.

I can't speak to exactly what the best way is to address this problem, as this must be fixed on Amazon's end but whatever conversion method is used on the list Campaigns and Ad Groups endpoints should probably be implemented here as well.

**Comment by ncoughlin** [comment]
Upon further examination I also believe that the keywordId's being returned are also being incorrectly rounded. For example if you make a text match request it appears that the vast majority of the keywordId's being returned are simply invalid. Same bigint rounding error.

**Comment by elizafarley** [comment]
Hi @ncoughlin, thanks for pointing this out! Would you mind submitting this as a bug report in [JIRA](https://amzn-clicks.atlassian.net/servicedesk/customer/portal/2/group/2/create/6) so our engineering team can investigate? In the JIRA ticket it would be helpful if you can include the relevant request IDs where you are seeing this behavior.

**Reply by ncoughlin** [reply]
I'll do that. Thanks Eliza.

**Comment by AdsAPIPM** [comment]
@ncoughlin would be good to share here how you addressed this issue.

**Reply by ncoughlin** [reply, marked as answer]
Yes gladly. So this was not a bug with the API, the numbers were being incorrectly rounded on my end by Axios before being returned.

Just a quick bit of background to give this some context. In JavaScript there is something called the "Max Safe Integer" which is the number 9007199254740991. You can read more about why that is the number here: [MDN: Number.MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)

Any attempt to perform mathematical operations on numbers larger than this will give inaccurate answers (the numbers will be rounded). Additionally if you attempt to convert one of these unsafe numbers from a string to a number, you will also get a rounding error.

![image](https://github.com/amzn/ads-advanced-tools-docs/assets/3167700/76500a6e-9763-4db0-80e8-6840b9bd74c4)

Lastly we need to understand that some campaign/adgroup/keyword id's, which are all numbers, are larger than this max safe number. Not all, but some.

So I was encountering an issue where I was getting back rounded id's for all the numbers that were larger than the max safe integer on various endpoints.

My API requests are made in Lambda using the Axios package. I logged the Axios response and saw that the id's were rounded there, and incorrectly assumed that they must be coming in from the Ads API that way. They weren't. Axios was transforming the result before returning it in my functions. Making a similar request in Postman confirmed that the id numbers were coming in correctly.

**SOLUTION:**

Axios has a `transformResponse` method that can be added to a request config that we can use in conjunction with a package called  [json-bigint](https://www.npmjs.com/package/json-bigint) to ensure that the "unsafe" id numbers are not rounded before being returned in the response.

In practice this is implemented like so:

```
config = {
          method: "get",
          url: `https://advertising-api.amazon.com/sb/keywords${queries}`,
          headers: {
            Accept: "application/vnd.sbkeyword.v3.2+json",
            Authorization: ACCESS_TOKEN,
            "Amazon-Advertising-API-ClientId": CLIENT_ID,
            "Amazon-Advertising-API-Scope": PROFILE_ID,
          },
          transformResponse: [
            (data) => {
              // Use json-bigint to parse the response data
              return JSONbig.parse(data);
            },
          ],
        }

let response = await axios(config);
```

I would recommend that this method (or similar) is used on any requests being made to API version 2 or 3 endpoints.
API version 4 endpoints return ID's as strings which solves this problem entirely.

Thank you Ads API Support team!
